Create pointer.js

markvantilburg opened the pull request:
A quick port of mouse.js to pointer events

markvantilburg committed:
Create pointer.js

A quick port of mouse.js to pointer events

markvantilburg committed:
Update pointer.js

missed a mouse

mgol reviewed (commented):
Thanks for the PR! Did you test the module with mouse & touch?

markvantilburg commented:
I tested it with a mouse (on desktop) and in chrome on my android, by replacing 'mouse' with 'pointer' with Slider and Sortable.

I tested with this file/js combination. More testing is needed for this feature.

[jquery-ui.min.1.14.1.js](https://github.com/user-attachments/files/22513914/jquery-ui.min.1.14.1.js)
[sort.html](https://github.com/user-attachments/files/22513915/sort.html)

markvantilburg commented:
@mgol is that safari code still needed?

mgol reviewed (changes requested):
I haven't tested this thoroughly, but the basic functionality works for the slider on both desktop & mobile with this PR + gh-2373; thanks for preparing these PRs!

That said, I think there are more changes to be made here:

1. Mouse & touch events have a different strategy for `*up` events - for mouse, they fire at the element above which the cursor was placed when it was relieved; for touch events, it fired at the element where the touch started. Pointer Events have a mechanism to opt into this touch behavior which is more convenient for drag operations; see MDN docs for [`setPointerCapture`](https://developer.mozilla.org/en-US/docs/Web/API/Element/setPointerCapture) & [`releasePointerCapture`](https://developer.mozilla.org/en-US/docs/Web/API/Element/releasePointerCapture).
2. Pointer Events support multi-touch. Ideally, we'd allow to drag both slider handles for the range slider, but perhaps that'd be too big of a change. In that case, we should at least ensure the second & subsequent pointers are ignored.
3. We need to handle the [`pointercancel` event](https://developer.mozilla.org/en-US/docs/Web/API/Element/pointercancel_event).

That's it for now. Would you like to continue the work on this PR?

markvantilburg committed:
Update pointer.js

Remove "ignoreMissingButtons" boolean and change the if to remove the custom safari code

markvantilburg committed:
Update pointer.js

In _pointerInit, the code used this.started = false, but everywhere else the code uses this._pointerStarted

markvantilburg committed:
Implement pointercancel event handling

Added pointercancel event handling to pointer widget.

markvantilburg commented:
Added pointercancel support, most of it should be implemented in the widgets

markvantilburg committed:
Refactor pointer event handling and cleanup

markvantilburg committed:
Rename pointer methods for consistency

_handlePointerCancel is the private event handler

_pointerCancel is the placeholder as the override point for subwidgets

markvantilburg committed:
Using isPrimary to Filter Pointer Events

Make sure only the first pointer will be handled

markvantilburg commented:
Only support the "first pointer" now
- Using isPrimary to Filter Pointer Events

markvantilburg committed:
Fix jsLint errors

Fix linting errors

markvantilburg committed:
Clean up pointer.js by removing blank lines

Removed unnecessary blank lines in pointer.js.

markvantilburg committed:
Refactor pointer event handling and capture logic

capture on the element and not on document

markvantilburg commented:
@mgol would it be possible to create a complete build that uses this instead of mouse in all the widgets? Just case sensitive replace "mouse/Mouse" with "pointer/Pointer" so it can be tested on multiple devices.
